### PR TITLE
fix list index out of range error for soda dataset

### PR DESCRIPTION
**Pull request by CloseChoice**
- fixed "index out of range" error in soda dataset
- simplified the logic in the dataset preprocessing
- added some plausibility checks
@theblackcat102 I am pinging you here directly, since you created it. Would be good to get your feedback on this.

### CODE REVIEW

File: model/model_training/custom_datasets/qa_datasets.py [hunk 1]
```diff
@@ -211,48 +212,57 @@ def __getitem__(self, index):
...
+            pairs = self._partition_list(data["dialogue"], 2)
+            # add dialog background to first pair.
+            # [[Q1, A1], [Q2, A2]] -> [[B + Q1, A1], [Q2, A2]]
+            pairs[0][0] = f"{dialogue_bg} {pairs[0][0]}"
+            # add previous history to conversation pairs
+            # e.g. [[B + Q1, A1], [Q2, A2]] -> [[B + Q1, A1], [B + Q1, A1, Q2, A2]]
+            pairs = list(accumulate(pairs, lambda x, y: x + y))
+            return pairs
 
     def __init__(self, cache_dir, input_max_length=1024) -> None:
```
Comment (andreaskoepf) [inline model/model_training/custom_datasets/qa_datasets.py:258]: Is `input_max_length` used?

Comment (CloseChoice) [reply model/model_training/custom_datasets/qa_datasets.py:258]: nope, just in [commented out code](https://github.com/LAION-AI/Open-Assistant/pull/2326/files#diff-c2caae63856734e93f51b5cbcb1695605d53a69538b8ea0d45471e2c94f95ee9L257). But I guess @theblackcat102 commented that out

Comment (theblackcat102) [reply model/model_training/custom_datasets/qa_datasets.py:258]: Yeah, we could, but the problem is how to pass in the parameters?

**Review by theblackcat102 — COMMENTED** [review]
LGTM

File: model/model_training/custom_datasets/qa_datasets.py (hunk 1 above)
Comment (CloseChoice) [reply model/model_training/custom_datasets/qa_datasets.py:268]: Added the option to pass the parameter right from the config. Using `**kwargs` should IMO be a general pattern that we could use more often for our datasets

**Review by andreaskoepf — APPROVED** [review]
Nice, thanks a lot for adapting to latest sft & rl loader format!

**Pull request merged**